### PR TITLE
implement ESSearchDAO multiple filter for metadata result

### PR DESCRIPTION
**Pull request by ybz1013**
## Summary
Implement multiple filters (with AND logic) for extracting metadata result 

## Context
The old implementation doesn't support multiple filters for extracting metadata result. If multiple sub-aggregations were present, the current logic would only pick up the last one, and the metadata result will be included as long as the picked one isn't empty (has doc count > 0).

## Testing Done
added unit test
./gradlew build

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/search/ESSearchDAO.java [hunk 1]
```diff
@@ -503,24 +504,41 @@ SearchResultMetadata extractSearchResultMetadata(@Nonnull SearchResponse searchR
...
     for (Terms.Bucket bucket : bucketList) {
       String key = bucket.getKeyAsString();
-      ParsedFilter parsedFilter = extractBucketAggregations(bucket);
-      // Gets filtered sub aggregation doc count if exist
-      Long docCount = parsedFilter != null ? parsedFilter.getDocCount() : bucket.getDocCount();
-      if (docCount > 0) {
-        aggResult.put(key, docCount);
+      List<ParsedFilter> parsedFilters = extractBucketAggregations(bucket);
+
+      assert parsedFilters != null;
```
Comment (zhixuanjia) [inline dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/search/ESSearchDAO.java:523]: this assert is not needed? The parsedFilter won't be null as a new list is always created at line 557?

File: dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/search/ESSearchDAO.java [hunk 2]
```diff
@@ -531,19 +549,21 @@ private Map<String, Long> extractTermAggregations(@Nonnull ParsedTerms terms) {
    * Extracts sub aggregations from one term bucket.
    *
    * @param bucket a term bucket
-   * @return a parsed filter if exist
+   * @return a list of parsed filter if exist
    */
   @Nullable
```
Comment (zhixuanjia) [inline dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/search/ESSearchDAO.java:554]: Should be @Nonnull instead? A new list is always instantiated.

File: dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/search/ESSearchDAO.java [hunk 3]
```diff
@@ -503,24 +504,41 @@ SearchResultMetadata extractSearchResultMetadata(@Nonnull SearchResponse searchR
...
+        // logic of these filters should be AND, so if any sub filter has doc document equals to 0, this bucket should return 0 results
+        for (ParsedFilter parsedFilter: parsedFilters) {
+          if (parsedFilter != null) {
+            minDocCount = Math.min(minDocCount, parsedFilter.getDocCount());
+          }
+        }
+
+        if (minDocCount > 0 && minDocCount != Long.MAX_VALUE) {
+          aggResult.put(key, minDocCount);
+        }
```
Comment (zhixuanjia) [inline dao-impl/elasticsearch-dao-7/src/main/java/com/linkedin/metadata/dao/search/ESSearchDAO.java:541]: Can you please explain a bit why we're finding the min doc count among a list of parsed filers?

**Review by zhixuanjia — COMMENTED** [review]
Can you add a break point at line 218 to see what is search request being sent to Elasticsearch? And add the search request payload to the PR description?

**Comment by ybz1013** [comment]
pause on this feature due to design discussion

**Comment by zhixuanjia** [comment]
new PR https://github.com/linkedin/datahub-gma/pull/339. Close this one.

**Pull request closed without merging**